File: mobile/run_e2e_tests.js
#!/usr/bin/env node
/**
 * Mobile End-to-End Test Runner
 * Executes comprehensive mobile end-to-end tests covering the complete user workflow
 * from video capture through upload, challenge creation, and playback
 */

const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

// ANSI color codes for console output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
};

function colorize(text, color) {
  return `${colors[color]}${text}${colors.reset}`;
}

function runCommand(command, args, description, options = {}) {
  return new Promise((resolve) => {
    console.log('\n' + '='.repeat(60));
    console.log(colorize(`Running: ${description}`, 'cyan'));
    console.log(colorize(`Command: ${command} ${args.join(' ')}`, 'blue'));
    console.log('='.repeat(60));

    const startTime = Date.now();
    const child = spawn(command, args, {
      stdio: 'inherit',
      cwd: options.cwd || process.cwd(),
      env: { ...process.env, ...options.env },
    });

    // Set timeout
    const timeout = setTimeout(() => {
      child.kill('SIGTERM');
      console.log(colorize(`⏰ TIMEOUT: ${description} took longer than 5 minutes`, 'yellow'));
      resolve(false);
    }, 300000); // 5 minutes

    child.on('close', (code) => {
      clearTimeout(timeout);
      const duration = ((Date.now() - startTime) / 1000).toFixed(2);
      
      if (code === 0) {
        console.log(colorize(`✅ SUCCESS: ${description} (${duration}s)`, 'green'));
        resolve(true);
      } else {
        console.log(colorize(`❌ FAILED: ${description} (${duration}s) - Exit code: ${code}`, 'red'));
        resolve(false);
      }
    });

    child.on('error', (error) => {
      clearTimeout(timeout);
      console.log(colorize(`💥 ERROR: ${description} failed with exception: ${error.message}`, 'red'));
      resolve(false);
    });
  });
}

async function checkDependencies() {
  console.log(colorize('🔍 Checking dependencies...', 'cyan'));
  
  // Check if package.json exists
  if (!fs.existsSync('package.json')) {
    console.log(colorize('❌ package.json not found', 'red'));
    return false;
  }

  // Check if node_modules exists
  if (!fs.existsSync('node_modules')) {
    console.log(colorize('⚠️  node_modules not found, installing dependencies...', 'yellow'));
    return await runCommand('npm', ['install'], 'Installing dependencies');
  }

  console.log(colorize('✅ Dependencies check passed', 'green'));
  return true;
}

async function main() {
  console.log(colorize('🚀 Starting Mobile End-to-End Test Suite', 'bright'));
  console.log(colorize('Testing complete mobile workflow from video capture to challenge playback', 'cyan'));

  const testResults = [];

  // Check dependencies
  const depsSuccess = await checkDependencies();
  testResults.push(['Dependencies Check', depsSuccess]);

  if (!depsSuccess) {
    console.log(colorize('❌ Cannot proceed without dependencies', 'red'));
    return 1;
  }

  // Core E2E Tests
  console.log(colorize('\n📱 Running Core Mobile E2E Tests', 'magenta'));

  // Complete E2E Workflow Tests
  const completeE2ESuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=CompleteE2EWorkflow.test.tsx', '--verbose', '--no-cache'],
    'Complete E2E Workflow Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Complete E2E Workflow', completeE2ESuccess]);

  // Existing End-to-End Integration Tests
  const existingE2ESuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=EndToEndIntegration.test.tsx', '--verbose', '--no-cache'],
    'Existing E2E Integration Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Existing E2E Integration', existingE2ESuccess]);

  // Component Integration Tests
  console.log(colorize('\n🧩 Running Component Integration Tests', 'magenta'));

  // Mobile Camera Integration Tests
  const cameraIntegrationSuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=MobileCameraIntegration.test.tsx', '--verbose', '--no-cache'],
    'Mobile Camera Integration Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Mobile Camera Integration', cameraIntegrationSuccess]);

  // Upload Integration Tests
  const uploadIntegrationSuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=UploadIntegrationComprehensive.test.tsx', '--verbose', '--no-cache'],
    'Upload Integration Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Upload Integration', uploadIntegrationSuccess]);

  // Mobile Redux Integration Tests
  const reduxIntegrationSuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=MobileReduxIntegration.test.tsx', '--verbose', '--no-cache'],
    'Mobile Redux Integration Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Mobile Redux Integration', reduxIntegrationSuccess]);

  // Cross-Device Tests
  console.log(colorize('\n📲 Running Cross-Device Tests', 'magenta'));

  // Cross-Device Media Accessibility Tests
  const crossDeviceSuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=CrossDeviceMediaAccessibility.test.tsx', '--verbose', '--no-cache'],
    'Cross-Device Media Accessibility Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Cross-Device Accessibility', crossDeviceSuccess]);

  // Device-Specific Tests
  console.log(colorize('\n🔧 Running Device-Specific Tests', 'magenta'));

  // Device Camera Recording Flow Tests
  const deviceCameraSuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=DeviceCameraRecordingFlow.test.tsx', '--verbose', '--no-cache'],
    'Device Camera Recording Flow Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Device Camera Recording', deviceCameraSuccess]);

  // Device End-to-End Camera Flow Tests
  const deviceE2ECameraSuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=DeviceEndToEndCameraFlow.test.tsx', '--verbose', '--no-cache'],
    'Device End-to-End Camera Flow Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Device E2E Camera Flow', deviceE2ECameraSuccess]);

  // Error Handling Tests
  console.log(colorize('\n⚠️  Running Error Handling Tests', 'magenta'));

  // Error Handling Tests
  const errorHandlingSuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=ErrorHandling.test.tsx', '--verbose', '--no-cache'],
    'Error Handling Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Error Handling', errorHandlingSuccess]);

  // Upload Error Handling Tests
  const uploadErrorSuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=UploadErrorHandlingComprehensive.test.ts', '--verbose', '--no-cache'],
    'Upload Error Handling Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Upload Error Handling', uploadErrorSuccess]);

  // Performance Tests
  console.log(colorize('\n⚡ Running Performance Tests', 'magenta'));

  // Mobile Media Capture Scenarios Tests
  const mediaCaptureSuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=MobileMediaCaptureScenarios.test.tsx', '--verbose', '--no-cache'],
    'Mobile Media Capture Scenarios Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Mobile Media Capture Scenarios', mediaCaptureSuccess]);

  // Permission Flow Tests
  const permissionFlowSuccess = await runCommand(
    'npm',
    ['test', '--', '--testPathPattern=PermissionFlows.test.tsx', '--verbose', '--no-cache'],
    'Permission Flow Tests',
    { env: { CI: 'true' } }
  );
  testResults.push(['Permission Flows', permissionFlowSuccess]);

  // Generate Test Coverage Report
  console.log(colorize('\n📊 Generating Test Coverage Report', 'magenta'));

  const coverageSuccess = await runCommand(
    'npm',
    ['run', 'test:coverage', '--', '--testPathPattern=.*E2E.*', '--verbose'],
    'Test Coverage Report Generation',
    { env: { CI: 'true' } }
  );
  testResults.push(['Test Coverage Report', coverageSuccess]);

  // Generate Test Results Summary
  console.log(colorize('\n📊 Test Results Summary', 'bright'));
  console.log('='.repeat(80));

  const totalTests = testResults.length;
  const passedTests = testResults.filter(([, success]) => success).length;
  const failedTests = totalTests - passedTests;

  testResults.forEach(([testName, success]) => {
    const status = success ? colorize('✅ PASS', 'green') : colorize('❌ FAIL', 'red');
    console.log(`${status.padEnd(20)} ${testName}`);
  });

  console.log('='.repeat(80));
  console.log(colorize(`Total Tests: ${totalTests}`, 'cyan'));
  console.log(colorize(`Passed: ${passedTests}`, 'green'));
  console.log(colorize(`Failed: ${failedTests}`, failedTests > 0 ? 'red' : 'green'));
  console.log(colorize(`Success Rate: ${((passedTests / totalTests) * 100).toFixed(1)}%`, 'cyan'));

  // Final status
  if (failedTests === 0) {
    console.log(colorize('\n🎉 ALL TESTS PASSED! Mobile end-to-end workflow is working correctly.', 'green'));
    return 0;
  } else {
    console.log(colorize(`\n⚠️  ${failedTests} TEST(S) FAILED. Please review the failures above.`, 'yellow'));
    return 1;
  }
}

// Handle process termination
process.on('SIGINT', () => {
  console.log(colorize('\n🛑 Test execution interrupted by user', 'yellow'));
  process.exit(1);
});

process.on('SIGTERM', () => {
  console.log(colorize('\n🛑 Test execution terminated', 'yellow'));
  process.exit(1);
});

// Run the main function
main()
  .then((exitCode) => {
    process.exit(exitCode);
  })
  .catch((error) => {
    console.error(colorize(`💥 Unexpected error: ${error.message}`, 'red'));
    console.error(error.stack);
    process.exit(1);
  });

File: backend/run_e2e_tests.py
#!/usr/bin/env python3
"""
End-to-End Test Runner
Executes comprehensive end-to-end tests covering the complete user workflow
from video capture through upload, merge, challenge creation, and playback
"""
import subprocess
import sys
import os
import time
from pathlib import Path

def run_command(command, description, cwd=None):
    """Run a command and return success status"""
    print(f"\n{'='*60}")
    print(f"Running: {description}")
    print(f"Command: {command}")
    print(f"{'='*60}")
    
    try:
        result = subprocess.run(
            command,
            shell=True,
            cwd=cwd,
            capture_output=True,
            text=True,
            timeout=300  # 5 minute timeout
        )
        
        if result.returncode == 0:
            print(f"✅ SUCCESS: {description}")
            if result.stdout:
                print("STDOUT:")
                print(result.stdout)
        else:
            print(f"❌ FAILED: {description}")
            print("STDERR:")
            print(result.stderr)
            if result.stdout:
                print("STDOUT:")
                print(result.stdout)
        
        return result.returncode == 0
    
    except subprocess.TimeoutExpired:
        print(f"⏰ TIMEOUT: {description} took longer than 5 minutes")
        return False
    except Exception as e:
        print(f"💥 ERROR: {description} failed with exception: {e}")
        return False

def main():
    """Main test runner function"""
    print("🚀 Starting End-to-End Test Suite")
    print("Testing complete workflow from video capture to challenge playback")
    
    # Get project root directory
    project_root = Path(__file__).parent.parent
    backend_dir = project_root / "backend"
    mobile_dir = project_root / "mobile"
    
    # Track test results
    test_results = []
    
    # Backend E2E Tests
    print("\n🔧 Running Backend End-to-End Tests")
    
    # Install backend test dependencies
    backend_deps_success = run_command(
        "pip install -r test-requirements.txt",
        "Installing backend test dependencies",
        cwd=backend_dir
    )
    test_results.append(("Backend Dependencies", backend_deps_success))
    
    if backend_deps_success:
        # Run complete E2E workflow tests
        backend_e2e_success = run_command(
            "python -m pytest tests/test_complete_e2e_workflow.py -v --tb=short",
            "Backend Complete E2E Workflow Tests",
            cwd=backend_dir
        )
        test_results.append(("Backend E2E Workflow", backend_e2e_success))
        
        # Run multi-video integration tests
        backend_multi_video_success = run_command(
            "python -m pytest tests/test_multi_video_upload_merge_integration.py -v --tb=short",
            "Backend Multi-Video Integration Tests",
            cwd=backend_dir
        )
        test_results.append(("Backend Multi-Video Integration", backend_multi_video_success))
        
        # Run challenge video API tests
        backend_api_success = run_command(
            "python -m pytest tests/test_challenge_video_api.py -v --tb=short",
            "Backend Challenge Video API Tests",
            cwd=backend_dir
        )
        test_results.append(("Backend Challenge Video API", backend_api_success))
        
        # Run video merge service tests
        backend_merge_success = run_command(
            "python -m pytest tests/test_video_merge_service.py -v --tb=short",
            "Backend Video Merge Service Tests",
            cwd=backend_dir
        )
        test_results.append(("Backend Video Merge Service", backend_merge_success))
        
        # Run existing end-to-end integration tests
        backend_existing_e2e_success = run_command(
            "python -m pytest tests/test_end_to_end_integration.py -v --tb=short",
            "Backend Existing E2E Integration Tests",
            cwd=backend_dir
        )
        test_results.append(("Backend Existing E2E Integration", backend_existing_e2e_success))
    
    # Mobile E2E Tests
    print("\n📱 Running Mobile End-to-End Tests")
    
    # Check if mobile directory exists and has package.json
    if mobile_dir.exists() and (mobile_dir / "package.json").exists():
        # Install mobile test dependencies
        mobile_deps_success = run_command(
            "npm install",
            "Installing mobile test dependencies",
            cwd=mobile_dir
        )
        test_results.append(("Mobile Dependencies", mobile_deps_success))
        
        if mobile_deps_success:
            # Run complete E2E workflow tests
            mobile_e2e_success = run_command(
                "npm test -- --testPathPattern=CompleteE2EWorkflow.test.tsx --verbose",
                "Mobile Complete E2E Workflow Tests",
                cwd=mobile_dir
            )
            test_results.append(("Mobile E2E Workflow", mobile_e2e_success))
            
            # Run existing end-to-end integration tests
            mobile_existing_e2e_success = run_command(
                "npm test -- --testPathPattern=EndToEndIntegration.test.tsx --verbose",
                "Mobile Existing E2E Integration Tests",
                cwd=mobile_dir
            )
            test_results.append(("Mobile Existing E2E Integration", mobile_existing_e2e_success))
            
            # Run mobile camera integration tests
            mobile_camera_success = run_command(
                "npm test -- --testPathPattern=MobileCameraIntegration.test.tsx --verbose",
                "Mobile Camera Integration Tests",
                cwd=mobile_dir
            )
            test_results.append(("Mobile Camera Integration", mobile_camera_success))
            
            # Run upload integration tests
            mobile_upload_success = run_command(
                "npm test -- --testPathPattern=UploadIntegrationComprehensive.test.tsx --verbose",
                "Mobile Upload Integration Tests",
                cwd=mobile_dir
            )
            test_results.append(("Mobile Upload Integration", mobile_upload_success))
    else:
        print("⚠️  Mobile directory not found or missing package.json, skipping mobile tests")
        test_results.append(("Mobile Tests", False))
    
    # Integration Tests (Backend + Mobile)
    print("\n🔗 Running Cross-Platform Integration Tests")
    
    # Run tests that verify backend-mobile integration
    if backend_deps_success:
        integration_success = run_command(
            "python -m pytest tests/ -k 'integration' -v --tb=short",
            "Cross-Platform Integration Tests",
            cwd=backend_dir
        )
        test_results.append(("Cross-Platform Integration", integration_success))
    
    # Performance and Load Tests
    print("\n⚡ Running Performance Tests")
    
    if backend_deps_success:
        # Run concurrent upload tests
        concurrent_success = run_command(
            "python -m pytest tests/test_complete_e2e_workflow.py::TestCompleteE2EWorkflow::test_concurrent_multi_user_upload_and_merge -v",
            "Concurrent Multi-User Upload Tests",
            cwd=backend_dir
        )
        test_results.append(("Concurrent Upload Tests", concurrent_success))
        
        # Run upload failure recovery tests
        recovery_success = run_command(
            "python -m pytest tests/test_complete_e2e_workflow.py::TestCompleteE2EWorkflow::test_upload_failure_recovery_and_resume -v",
            "Upload Failure Recovery Tests",
            cwd=backend_dir
        )
        test_results.append(("Upload Recovery Tests", recovery_success))
    
    # Generate Test Report
    print("\n📊 Test Results Summary")
    print("="*80)
    
    total_tests = len(test_results)
    passed_tests = sum(1 for _, success in test_results if success)
    failed_tests = total_tests - passed_tests
    
    for test_name, success in test_results:
        status = "✅ PASS" if success else "❌ FAIL"
        print(f"{status:<10} {test_name}")
    
    print("="*80)
    print(f"Total Tests: {total_tests}")
    print(f"Passed: {passed_tests}")
    print(f"Failed: {failed_tests}")
    print(f"Success Rate: {(passed_tests/total_tests)*100:.1f}%")
    
    # Final status
    if failed_tests == 0:
        print("\n🎉 ALL TESTS PASSED! End-to-end workflow is working correctly.")
        return 0
    else:
        print(f"\n⚠️  {failed_tests} TEST(S) FAILED. Please review the failures above.")
        return 1

if __name__ == "__main__":
    exit_code = main()
    sys.exit(exit_code)